crates/ruff_macros/src/combine.rs

```diff
@@ -0,0 +1,43 @@
+use quote::{quote, quote_spanned};
+use syn::{Data, DataStruct, DeriveInput, Fields};
+
+pub(crate) fn derive_impl(input: DeriveInput) -> syn::Result<proc_macro2::TokenStream> {
+    let DeriveInput { ident, data, .. } = input;
+
+    match data {
+        Data::Struct(DataStruct {
+            fields: Fields::Named(fields),
+            ..
+        }) => {
+            let output: Vec<_> = fields
+                .named
+                .iter()
+                .map(|field| {
+                    let ident = field
+                        .ident
+                        .as_ref()
+                        .expect("Expected to handle named fields");
+
+                    quote_spanned!(
+                        ident.span() => crate::project::combine::Combine::combine_with(&mut self.#ident, other.#ident)
+                    )
+                })
+                .collect();
+
+            Ok(quote! {
+                #[automatically_derived]
+                impl crate::project::combine::Combine for #ident {
+                    fn combine_with(&mut self, other: Self) {
+                        #(
+                            #output
+                        );*
+                    }
+                }
+            })
+        }
+        _ => Err(syn::Error::new(
+            ident.span(),
+            "Can only derive Combine from structs with named fields.",
+        )),
```

Comment:
Mostly just curious: Would it make sense (and is it possible?) to auto-derive `Combine` for enums by simply taking the value from the side which takes precendence and forgetting the other one?

Not sure if that's what we want in all cases though. In the `Option` case for example, we probably ignore overriding with `None`.

Reply:
Neither ruff nor uv required deriving `Combine` for enums and I'm not 100% sure what the semantics should be for enum with fields (we may want to merge them?). 

I suggest that we defer this until we have a use case where we can discuss the desired semantics.